### PR TITLE
Match V8 semantics for CallSite.getTypeName/getFunctionName/getMethodName

### PR DESCRIPTION
**Pull request by robobun**
## Problem

Reported in #30938: `async_hooks` + `process._fatalException` + `tap` produces TAP subtest errors formatted as `undefined.<anonymous>` instead of the receiver's class name.

```
# Bun 1.3.14
stack: |
  undefined.<anonymous> (fail.cjs:29:12)
at:
  function: undefined.<anonymous>
```

```
# Node 24
stack: |
  Test.<anonymous> (fail.cjs:29:9)
at:
  function: Test.<anonymous>
```

The literal string `"undefined"` appears because `source-map-support` does `typeName + "."` when formatting method-call frames, and Bun's `CallSite.getTypeName()` was returning `typeof this` (so `"undefined"`, `"object"`, `"function"`) where V8/Node returns the receiver's constructor name — or `null` when the receiver can't be identified.

Two neighbouring methods had the same shape of mismatch:

- `getFunctionName()` returned `""` for anonymous frames where V8 returns `null`.
- `getMethodName()` delegated to `getFunctionName` (there's even a `// TODO` on it), so it reported the function's own name even for plain, non-method frames where V8 returns `null`.

## Fix

`src/jsc/bindings/CallSitePrototype.cpp`:

- `getTypeName` — null/undefined receiver → `null`; otherwise `JSObject::calculatedClassName`, which already implements V8's cascade (own `constructor` → `proto.constructor` → `@@toStringTag` → `[[Class]]` → `"Object"`). Global object → `null`.
- `getFunctionName` — empty stored name → `null`.
- `getMethodName` — `null`. V8 resolves the method name by matching the frame's function against the receiver's own/inherited properties, but Bun's captured frames don't retain the receiver (see the caveat), so the property can't be identified. `null` is also what V8 returns when it can't determine a method name, and it's strictly better than the old delegation that reported the function's own name for non-method frames.

## Caveat

Bun can't reproduce Node's `Test.<anonymous>` output because `JSC::StackFrame` doesn't persist the receiver through stack capture — by the time `Error.prepareStackTrace` runs the callframe is gone, so `CallSite.getThis()` is `undefined` for captured frames (verified across object methods, class methods, `.call(obj)`, and constructors). Consequently `getTypeName()` and `getMethodName()` return `null` in practice today; the receiver-dependent branches activate only if a future change stashes `this` on each frame.

Net effect for the issue: the TAP output goes from `undefined.<anonymous>` to `null.<anonymous>` (and `function: undefined.<anonymous>` to `function: null.<anonymous>`). Still not Node's `Test.<anonymous>`, but consistent with what V8 returns when the receiver is unavailable, and the `null` return value unblocks ecosystem libraries that guard with `typeName && typeName !== 'Object'`. Going the rest of the way needs the receiver captured on every frame — a separate, bigger change.

## Verification

`test/js/node/v8/capture-stack-trace.test.js`:

- Three new regressions pinned to #30938 assert `null` return for each method.
- Two existing sanity-check assertions were updated from the buggy values (`"undefined"`, `"f3"` for `getMethodName` of a top-level function) to the V8-correct `null` — these encoded the bug.

```
bun bd test test/js/node/v8/capture-stack-trace.test.js
 43 pass
 0 fail
```

Fail-before (`USE_SYSTEM_BUN=1`):

```
CallSite.p.getTypeName ... Expected: null / Received: "undefined"
CallSite.p.getFunctionName ... Expected: null / Received: ""
CallSite.p.getMethodName ... Expected: null / Received: "topLevelFn"
```

Related test files still green:

```
bun bd test test/js/node/v8/                                   → 46 pass
bun bd test test/regression/issue/fix-bindings-stack-trace.test.ts
         test/regression/issue/prepare-stack-trace-crash.test.ts
         test/regression/issue/circular-error-stack.test.ts
         test/regression/issue/23022-stack-trace-iterator.test.ts
         test/regression/issue/circular-error-stack-edge-cases.test.ts → 12 pass
```

Fixes #30938

### CODE REVIEW

File: test/js/node/v8/capture-stack-trace.test.js [hunk 1]
```diff
@@ -466,6 +469,52 @@ test("CallFrame.p.isConstructor", () => {
...
+    e.stack;
+  }
+  Error.prepareStackTrace = (_, s) => {
+    // topLevelFn is a plain function call — it's not stored at a property of
+    // `this`, so V8 returns null. Previously Bun delegated to getFunctionName.
+    expect(s[0].getMethodName()).toBe(null);
+  };
+  topLevelFn();
+  Error.prepareStackTrace = prevPrepareStackTrace;
+});
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/node/v8/capture-stack-trace.test.js]: Done in e0dccb823a: added expect.hasAssertions() to all three tests and wrapped the stack access in try/finally. Note the file already has a suite-level afterEach that resets Error.prepareStackTrace to the original, so later tests were not actually at risk of poisoning, but the local try/finally and hasAssertions make each test self-contained. Left the pre-existing f3/sanity-check test to the afterEach to keep the diff focused.

File: src/jsc/bindings/CallSitePrototype.cpp [hunk 2]
```diff
@@ -113,16 +139,32 @@ JSC_DEFINE_HOST_FUNCTION(callSiteProtoFuncGetFunction, (JSGlobalObject * globalO
...
+// identified, and we return null. This also matches what V8 returns when it
+// can't determine a method name (e.g. plain function calls). Previously this
+// delegated to getFunctionName, which reported the function's own name even
+// for non-method frames.
 JSC_DEFINE_HOST_FUNCTION(callSiteProtoFuncGetMethodName, (JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
 {
-    return callSiteProtoFuncGetFunctionName(globalObject, callFrame);
+    ENTER_PROTO_FUNC();
+    return JSC::JSValue::encode(JSC::jsNull());
 }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/CallSitePrototype.cpp]: Trimmed both comments to the 3-line limit in 767a3b1.

On the accuracy point: `thisValue()` really is `undefined` for the frames that reach `Error.prepareStackTrace`, including non-strict ones. Those frames are built from captured `JSC::StackFrame` objects via `JSCStackFrame(VM&, const JSC::StackFrame&)`, whose constructor sets `m_callFrame(nullptr)` (ErrorStackTrace.cpp:266). In `CallSite::finishCreation` the non-strict branch is `if (callFrame && callFrame->thisValue())` — with `callFrame == nullptr` it falls through to `m_thisValue = jsUndefined()`. I also verified empirically: `getThis()` returns `undefined` for object methods, class methods, `sloppyFn.call(obj)`, and constructors. The sanity-check test asserts `getThis() === undefined`, not a live receiver. So the comment's rationale (receiver not retained) is correct; implementing the property-walk wouldn't help without the receiver. Kept the wording but shortened it.

File: src/jsc/bindings/CallSitePrototype.cpp [hunk 3]
```diff
@@ -100,11 +100,32 @@ JSC_DEFINE_HOST_FUNCTION(callSiteProtoFuncGetThis, (JSGlobalObject * globalObjec
     return JSC::JSValue::encode(callSite->thisValue());
 }
 
-// TODO: doesn't get class name
+// Matches V8's GetTypeName: the receiver's constructor name, else null for
+// null/undefined/global/strict receivers. Never returns the literal string
+// "undefined".
 JSC_DEFINE_HOST_FUNCTION(callSiteProtoFuncGetTypeName, (JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
 {
     ENTER_PROTO_FUNC();
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/CallSitePrototype.cpp:108]: Good catch. Updated the CallSite table in docs/runtime/debugger.mdx in 249f68c5: getTypeName -> "constructor name of the receiver, or null", getFunctionName -> "function name, or null if anonymous", getMethodName -> "null (receiver not retained)".

File: test/js/node/v8/capture-stack-trace.test.js [hunk 4]
```diff
@@ -466,6 +469,65 @@ test("CallFrame.p.isConstructor", () => {
   Error.prepareStackTrace = prevPrepareStackTrace;
 });
 
+// Regression for https://github.com/oven-sh/bun/issues/30938
+// getTypeName/getFunctionName/getMethodName must match V8 (return null,
+// not the string "undefined" or the empty string). tap + source-map-support
+// concatenates `typeName + "."` into the formatted frame, so the old
+// behaviour rendered as `undefined.<anonymous>` in TAP error output.
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/node/v8/capture-stack-trace.test.js]: Done in 249f68c5. Reduced the block to the issue-URL line, and also dropped the bug-history clause from the inline comment in the getMethodName test. The test names carry the intent.
